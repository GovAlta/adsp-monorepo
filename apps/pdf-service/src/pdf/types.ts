import { AdspId } from '@abgov/adsp-service-sdk';
import * as NodeCache from 'node-cache';

export interface TemplateService {
  getTemplateFunction(template: string, channel?: string): (context: unknown) => string;
  setTenantToken(token: string): void;
  getTenantToken(): string;
  getFileServiceCache(): NodeCache;
  populateFileList(token: string, tenantIdValue: string): Promise<File[]>;
}

export interface PdfServiceProps {
  content: string;
  footer?: string;
  header?: string;
}

export interface File {
  id: string;
  recordId: string;
  filename: string;
  size: number;
  createdBy: UserInfo;
  created: Date;
  lastAccessed?: Date;
}

export interface UserInfo {
  id: string;
  name: string;
}



export interface PdfService {
  generatePdf({ content, footer, header }: PdfServiceProps): Promise<Buffer>;
}

export interface FileResult {
  urn: string;
  id: string;
  filename: string;
}

export interface FileService {
  typeExists(tenantId: AdspId, fileType: string): Promise<boolean>;
  upload(tenantId: AdspId, fileType: string, recordId: string, filename: string, content: Buffer): Promise<FileResult>;
}

export interface PdfTemplate {
  tenantId: AdspId;
  id: string;
  name: string;
  description: string;
  template: string;
  footer?: string;
  header?: string;
  additionalStyles?: string;
}

export type PdfJobStatus = 'queued' | 'completed' | 'failed';

export interface PdfJob {
  tenantId: AdspId;
  id: string;
  status: PdfJobStatus;
  result?: FileResult;
}
