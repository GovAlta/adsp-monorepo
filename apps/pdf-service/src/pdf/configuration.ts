export const configurationSchema = {
  type: 'object',
  additionalProperties: {
    type: 'object',
    properties: {
      id: {
        type: 'string',
        pattern: '^[a-zA-Z0-9-_ ]{1,50}$',
      },
      name: {
        type: 'string',
      },
      description: {
        type: 'string',
      },
      template: {
        type: 'string',
      },
    },
    required: ['id', 'name', 'template'],
  },
};

