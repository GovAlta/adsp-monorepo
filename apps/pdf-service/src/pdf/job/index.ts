import { WorkQueueService } from '@core-services/core-common';
import { createGenerateJob, GenerateJobProps } from './generate';
import { PdfServiceWorkItem } from './types';

export * from './types';

interface PdfJobProps extends GenerateJobProps {
  queueService: WorkQueueService<PdfServiceWorkItem>;
}

export const createPdfJobs = (props: PdfJobProps): void => {
  const generateJob = createGenerateJob(props);

  props.queueService.getItems().subscribe(({ item, retryOnError, done }) => {
    generateJob(item, retryOnError, done);
  });

  props.logger.info('Initialized PDF generate job.', { context: 'JobsFactory' });
};
