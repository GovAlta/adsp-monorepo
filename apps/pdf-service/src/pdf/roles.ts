export enum ServiceRoles {
  PdfGenerator = 'pdf-generator',
}
