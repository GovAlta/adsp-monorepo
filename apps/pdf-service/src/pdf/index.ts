import { AdspId, ConfigurationService, EventService, TokenProvider } from '@abgov/adsp-service-sdk';
import { WorkQueueService } from '@core-services/core-common';
import { Application } from 'express';
import { Logger } from 'winston';
import { createPdfJobs, PdfServiceWorkItem } from './job';
import { createPdfRouter } from './router';
import { FileService } from './types';

export * from './configuration';
export * from './events';
export * from './fileTypes';
export * from './roles';
export * from './types';
export * from './job';

interface MiddlewareProps {
  logger: Logger;
  serviceId: AdspId;
  tokenProvider: TokenProvider;
  configurationService: ConfigurationService;
  fileService: FileService;
  eventService: EventService;
  queueService: WorkQueueService<PdfServiceWorkItem>;
}

export function applyPdfMiddleware(app: Application, props: MiddlewareProps): Application {
  createPdfJobs(props);

  const router = createPdfRouter(props);
  app.use('/pdf/v1', router);

  return app;
}
