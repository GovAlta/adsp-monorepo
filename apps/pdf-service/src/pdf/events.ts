import { AdspId, DomainEvent, DomainEventDefinition, Stream } from '@abgov/adsp-service-sdk';
import { ServiceRoles } from './roles';
import { FileResult } from './types';

export const PDF_GENERATION_QUEUED = 'pdf-generation-queued';
export const PdfGenerationQueuedDefinition: DomainEventDefinition = {
  name: PDF_GENERATION_QUEUED,
  description: 'Signalled when a PDF generation job is queued.',
  payloadSchema: {
    type: 'object',
    properties: {
      jobId: { type: 'string' },
      templateId: { type: 'string' },
      requestedBy: {
        type: 'object',
        properties: {
          id: { type: 'string' },
          name: { type: 'string' },
        },
      },
    },
  },
};

export const PDF_GENERATED = 'pdf-generated';
export const PdfGeneratedDefinition: DomainEventDefinition = {
  name: PDF_GENERATED,
  description: 'Signalled when a PDF has been generated.',
  payloadSchema: {
    type: 'object',
    properties: {
      jobId: { type: 'string' },
      templateId: { type: 'string' },
      file: {
        type: 'object',
        properties: {
          id: { type: 'string' },
          urn: { type: 'string' },
          filename: { type: 'string' },
        },
      },
      requestedBy: {
        type: 'object',
        properties: {
          id: { type: 'string' },
          name: { type: 'string' },
        },
      },
    },
  },
  interval: {
    namespace: 'pdf-service',
    name: PDF_GENERATION_QUEUED,
    metric: ['pdf-service', 'pdf-generation'],
  },
};

export const PDF_GENERATION_FAILED = 'pdf-generation-failed';
export const PdfGenerationFailedDefinition: DomainEventDefinition = {
  name: PDF_GENERATION_FAILED,
  description: 'Signalled when a PDF generation has failed.',
  payloadSchema: {
    type: 'object',
    properties: {
      jobId: { type: 'string' },
      templateId: { type: 'string' },
      requestedBy: {
        type: 'object',
        properties: {
          id: { type: 'string' },
          name: { type: 'string' },
        },
      },
    },
  },
  interval: {
    namespace: 'pdf-service',
    name: PDF_GENERATION_QUEUED,
    metric: ['pdf-service', 'pdf-generation-failed'],
  },
};

export const PdfGenerationUpdatesStream: Stream = {
  id: 'pdf-generation-updates',
  name: 'PDF generation updates',
  description: 'Provides update events for PDF generation.',
  subscriberRoles: [`urn:ads:platform:pdf-service:${ServiceRoles.PdfGenerator}`],
  publicSubscribe: false,
  events: [
    {
      namespace: 'pdf-service',
      name: PDF_GENERATION_QUEUED,
    },
    {
      namespace: 'pdf-service',
      name: PDF_GENERATED,
    },
    {
      namespace: 'pdf-service',
      name: PDF_GENERATION_FAILED,
    },
  ],
};

export const pdfGenerationQueued = (
  tenantId: AdspId,
  jobId: string,
  templateId: string,
  requestedBy: { id: string; name: string }
): DomainEvent => ({
  name: PDF_GENERATION_QUEUED,
  tenantId,
  correlationId: jobId,
  context: { jobId, templateId },
  timestamp: new Date(),
  payload: {
    jobId,
    templateId,
    requestedBy,
  },
});

export const pdfGenerated = (
  tenantId: AdspId,
  jobId: string,
  templateId: string,
  file: FileResult,
  requestedBy: { id: string; name: string }
): DomainEvent => ({
  name: PDF_GENERATED,
  tenantId,
  correlationId: jobId,
  context: { jobId, templateId },
  timestamp: new Date(),
  payload: {
    jobId,
    templateId,
    file,
    requestedBy,
  },
});

export const pdfGenerationFailed = (
  tenantId: AdspId,
  jobId: string,
  templateId: string,
  requestedBy: { id: string; name: string }
): DomainEvent => ({
  name: PDF_GENERATION_FAILED,
  tenantId,
  correlationId: jobId,
  context: { jobId, templateId },
  timestamp: new Date(),
  payload: {
    jobId,
    templateId,
    requestedBy,
  },
});
