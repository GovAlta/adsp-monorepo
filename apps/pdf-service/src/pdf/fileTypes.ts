import { FileType } from '@abgov/adsp-service-sdk';
import { ServiceRoles } from './roles';

export const GENERATED_PDF = 'generated-pdf';
export const GeneratedPdfType: FileType = {
  id: GENERATED_PDF,
  name: 'Generated PDF',
  anonymousRead: false,
  securityClassification: 'Protected_A',
  readRoles: [`urn:ads:platform:pdf-service:${ServiceRoles.PdfGenerator}`],
  updateRoles: [`urn:ads:platform:tenant-service:platform-service`],
  rules: {
    retention: {
      active: true,
      createdAt: '2023-04-20T17:19:22Z',
      deleteInDays: 30,
    },
  },
};
