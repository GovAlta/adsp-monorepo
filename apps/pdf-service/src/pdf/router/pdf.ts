import { AdspId, EventService, isAllowedUser, UnauthorizedUserError } from '@abgov/adsp-service-sdk';
import {
  createValidationHandler,
  InvalidOperationError,
  NotFoundError,
  WorkQueueService,
} from '@core-services/core-common';
import { Request, RequestHandler, Response, Router } from 'express';
import { body, param } from 'express-validator';
import { Logger } from 'winston';
import { pdfGenerationQueued } from '../events';
import { GENERATED_PDF } from '../fileTypes';
import { PdfServiceWorkItem } from '../job';
import { PdfTemplateEntity } from '../model';
import { PdfJobRepository } from '../repository';
import { ServiceRoles } from '../roles';
import { FileService, PdfJob } from '../types';

export interface RouterProps {
  serviceId: AdspId;
  logger: Logger;
  repository: PdfJobRepository;
  queueService: WorkQueueService<PdfServiceWorkItem>;
  eventService: EventService;
  fileService: FileService;
}

function mapPdfTemplate({ id, name, description, template }: PdfTemplateEntity) {
  return {
    id,
    name,
    description,
    template,
  };
}

function mapJob(serviceId: AdspId, { id, status, result }: PdfJob) {
  return {
    urn: `${serviceId}:v1:/jobs/${id}`,
    id,
    status,
    result: result ? { urn: result.urn, id: result.id, filename: result.filename } : null,
  };
}

export const getTemplates: RequestHandler = async (req, res, next) => {
  try {
    const [configuration] = await req.getConfiguration<Record<string, PdfTemplateEntity>>();
    res.send(Object.values(configuration).map(mapPdfTemplate));
  } catch (err) {
    next(err);
  }
};

const TEMPLATE = 'template';
export function getTemplate(templateIn: 'params' | 'body'): RequestHandler {
  return async (req, res, next) => {
    try {
      const { templateId } = req[templateIn];
      const [configuration] = await req.getConfiguration<Record<string, PdfTemplateEntity>>();
      const template = configuration[templateId];

      if (!template) {
        throw new NotFoundError('PDF Template', templateId);
      }

      req[TEMPLATE] = template;

      next();
    } catch (err) {
      next(err);
    }
  };
}

export function generatePdf(
  serviceId: AdspId,
  repository: PdfJobRepository,
  eventService: EventService,
  fileService: FileService,
  queueService: WorkQueueService<PdfServiceWorkItem>
): RequestHandler {
  return async (req, res, next) => {
    try {
      const user = req.user;
      const tenantId = req.tenant.id;
      const { templateId, fileType, filename, recordId, data } = req.body;
      const template: PdfTemplateEntity = req[TEMPLATE];

      if (!isAllowedUser(user, template.tenantId, ServiceRoles.PdfGenerator)) {
        throw new UnauthorizedUserError('generate pdf', user);
      }

      if (fileType) {
        const typeExists = await fileService.typeExists(tenantId, fileType);
        if (!typeExists) {
          throw new InvalidOperationError(`Specified target file type '${fileType}' cannot be found.`);
        }
      }

      const job = await repository.create(tenantId);
      await queueService.enqueue({
        timestamp: new Date(),
        work: 'generate',
        jobId: job.id,
        tenantId: `${tenantId}`,
        fileType: fileType || GENERATED_PDF,
        templateId,
        filename,
        recordId: recordId || job.id,
        data: data || {},
        requestedBy: {
          id: user.id,
          name: user.name,
        },
      });

      eventService.send(pdfGenerationQueued(tenantId, job.id, templateId, { id: user.id, name: user.name }));

      res.send(mapJob(serviceId, job));
    } catch (err) {
      next(err);
    }
  };
}

export function getGeneratedFile(serviceId: AdspId, repository: PdfJobRepository): RequestHandler {
  return async (req, res, next) => {
    try {
      const user = req.user;

      const { jobId } = req.params;

      const job = await repository.get(jobId);
      if (!job) {
        throw new NotFoundError('Generated PDF', jobId);
      }

      if (!isAllowedUser(user, job.tenantId, ServiceRoles.PdfGenerator)) {
        throw new UnauthorizedUserError('get generated pdf', user);
      }

      res.send(mapJob(serviceId, job));
    } catch (err) {
      next(err);
    }
  };
}

export function createPdfRouter({
  serviceId,
  repository,
  eventService,
  fileService,
  queueService,
}: RouterProps): Router {
  const router = Router();

  router.get('/templates', getTemplates);
  router.get(
    '/templates/:templateId',
    createValidationHandler(param('templateId').isString().isLength({ min: 1, max: 50 })),
    getTemplate('params'),
    (req: Request, res: Response) => res.send(mapPdfTemplate(req[TEMPLATE]))
  );
  router.post(
    '/jobs',
    createValidationHandler(
      body('operation').isIn(['generate']),
      body('templateId').isString().isLength({ min: 1, max: 50 }),
      body('data').optional().isObject(),
      body('filename').isString().isLength({ min: 1, max: 50 }),
      body('fileType').optional().isString().isLength({ min: 1, max: 50 }),
      body('recordId').optional().isString()
    ),
    getTemplate('body'),
    generatePdf(serviceId, repository, eventService, fileService, queueService)
  );
  router.get('/jobs/:jobId', createValidationHandler(param('jobId').isUUID()), getGeneratedFile(serviceId, repository));

  return router;
}
