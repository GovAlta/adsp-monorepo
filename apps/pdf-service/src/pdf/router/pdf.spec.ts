import { adspId, UnauthorizedUserError } from '@abgov/adsp-service-sdk';
import { InvalidOperationError, NotFoundError } from '@core-services/core-common';
import { Request, Response } from 'express';
import { Logger } from 'winston';
import { PDF_GENERATION_QUEUED } from '../events';
import { ServiceRoles } from '../roles';
import { createPdfRouter, generatePdf, getGeneratedFile, getTemplate, getTemplates } from './pdf';
import axios from 'axios';

jest.mock('axios');
const axiosMock = axios as jest.Mocked<typeof axios>;

describe('pdf', () => {
  const serviceId = adspId`urn:ads:platform:pdf-service`;
  const tenantId = adspId`urn:ads:platform:tenant-service:v2:/tenants/test`;

  const loggerMock = {
    debug: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
  } as unknown as Logger;

  const repositoryMock = {
    create: jest.fn(),
    get: jest.fn(),
    update: jest.fn(),
  };

  const eventServiceMock = {
    send: jest.fn(),
  };

  const fileServiceMock = {
    typeExists: jest.fn(),
    upload: jest.fn(),
  };

  const queueServiceMock = {
    enqueue: jest.fn(),
    getItems: jest.fn(),
    isConnected: jest.fn(),
  };

  const configuration = {
    test: {
      id: 'test',
      name: 'Test template',
      description: 'This is a test.',
      template: '',
    },
  };

  beforeEach(() => {
    eventServiceMock.send.mockReset();
    queueServiceMock.enqueue.mockReset();
    repositoryMock.create.mockReset();
    repositoryMock.get.mockReset();
    fileServiceMock.typeExists.mockReset();
  });

  const serviceDirectoryMock = {
    getServiceUrl: jest.fn(() => Promise.resolve(new URL('http:/localhost:80'))),
    getResourceUrl: jest.fn(() => Promise.resolve(new URL('http:/localhost:80'))),
  };

  it('can create router', () => {
    const router = createPdfRouter({
      serviceId,
      logger: loggerMock,
      repository: repositoryMock,
      queueService: queueServiceMock,
      eventService: eventServiceMock,
      fileService: fileServiceMock,
      directory: serviceDirectoryMock,
    });

    expect(router).toBeTruthy();
  });

  describe('getTemplates', () => {
    it('can get templates', async () => {
      const req = {
        getConfiguration: jest.fn(),
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();
      req.getConfiguration.mockResolvedValueOnce([configuration]);
      await getTemplates(req as unknown as Request, res as unknown as Response, next);

      expect(res.send).toHaveBeenCalledWith(expect.arrayContaining([configuration.test]));
      expect(res.send.mock.calls[0][0]).toMatchSnapshot();
    });

    it('can call next for error', async () => {
      const req = {
        getConfiguration: jest.fn(),
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();
      req.getConfiguration.mockRejectedValueOnce(new Error('oh noes!'));
      await getTemplates(req as unknown as Request, res as unknown as Response, next);

      expect(next).toHaveBeenCalledWith(expect.any(Error));
      expect(res.send).not.toHaveBeenCalled();
    });
  });

  describe('getTemplate', () => {
    it('can create handler', () => {
      const handler = getTemplate('params');
      expect(handler).toBeTruthy();
    });

    it('can get template', async () => {
      const req = {
        params: {
          templateId: 'test',
        },
        getConfiguration: jest.fn(),
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();
      req.getConfiguration.mockResolvedValueOnce([configuration]);
      const handler = getTemplate('params');
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(next).toHaveBeenCalledWith();
      expect(req['template']).toBe(configuration.test);
    });

    it('can get template by body property', async () => {
      const req = {
        body: {
          templateId: 'test',
        },
        getConfiguration: jest.fn(),
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();
      req.getConfiguration.mockResolvedValueOnce([configuration]);
      const handler = getTemplate('body');
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(next).toHaveBeenCalledWith();
      expect(req['template']).toBe(configuration.test);
    });

    it('can call next with not found', async () => {
      const req = {
        params: {
          templateId: 'test',
        },
        getConfiguration: jest.fn(),
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();
      req.getConfiguration.mockResolvedValueOnce([{}]);
      const handler = getTemplate('params');
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(NotFoundError));
    });
  });

  describe('generatePdf', () => {
    it('can create handler', () => {
      const handler = generatePdf(
        serviceId,
        repositoryMock,
        eventServiceMock,
        fileServiceMock,
        queueServiceMock,
        loggerMock,
        serviceDirectoryMock
      );
      expect(handler).toBeTruthy();
    });

    it('can generate pdf', async () => {
      const req = {
        tenant: {
          id: tenantId,
        },
        user: {
          tenantId,
          id: 'test',
          name: 'tester',
          roles: [ServiceRoles.PdfGenerator],
        },
        body: {
          templateId: 'test',
          filename: 'test.pdf',
          data: {},
        },
        template: configuration.test,
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();

      repositoryMock.create.mockResolvedValueOnce({ id: 'job1' });
      fileServiceMock.typeExists.mockResolvedValueOnce(true);
      queueServiceMock.enqueue.mockResolvedValueOnce(null);
      const handler = generatePdf(
        serviceId,
        repositoryMock,
        eventServiceMock,
        fileServiceMock,
        queueServiceMock,
        loggerMock,
        serviceDirectoryMock
      );
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(next).not.toHaveBeenCalled();
      expect(queueServiceMock.enqueue).toHaveBeenCalledWith(
        expect.objectContaining({
          work: 'generate',
          jobId: 'job1',
          templateId: req.body.templateId,
          data: req.body.data,
          filename: req.body.filename,
          requestedBy: expect.objectContaining({ id: req.user.id, name: req.user.name }),
        })
      );
      expect(eventServiceMock.send).toHaveBeenCalledWith(
        expect.objectContaining({ tenantId, name: PDF_GENERATION_QUEUED })
      );
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ id: 'job1' }));
      expect(res.send.mock.calls[0][0]).toMatchSnapshot();
    });

    it('can generate core pdf', async () => {
      const req = {
        tenant: {
          id: tenantId,
        },
        user: {
          tenantId,
          id: 'test',
          name: 'tester',
          roles: [ServiceRoles.PdfGenerator],
          token: { bearer: 'abc' },
        },
        body: {
          templateId: 'test',
          filename: 'test.pdf',
          data: {},
          formId: 'test-form-id',
        },
        template: configuration.test,
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();

      repositoryMock.create.mockResolvedValueOnce({ id: 'job1' });
      fileServiceMock.typeExists.mockResolvedValueOnce(true);
      queueServiceMock.enqueue.mockResolvedValueOnce(null);

      axiosMock.get
        .mockResolvedValueOnce({
          data: {
            results: [{ id: 'formId' }],
          },
        })
        .mockResolvedValueOnce({ data: { firstName: 'bob', lastName: 'smith' } });

      const handler = generatePdf(
        serviceId,
        repositoryMock,
        eventServiceMock,
        fileServiceMock,
        queueServiceMock,
        loggerMock,
        serviceDirectoryMock
      );
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(next).not.toHaveBeenCalled();
      expect(queueServiceMock.enqueue).toHaveBeenCalledWith(
        expect.objectContaining({
          work: 'generate',
          jobId: 'job1',
          templateId: req.body.templateId,
          data: req.body.data,
          filename: req.body.filename,
          requestedBy: expect.objectContaining({ id: req.user.id, name: req.user.name }),
        })
      );
      expect(eventServiceMock.send).toHaveBeenCalledWith(
        expect.objectContaining({ tenantId, name: PDF_GENERATION_QUEUED })
      );
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ id: 'job1' }));
      expect(res.send.mock.calls[0][0]).toMatchSnapshot();
    });

    it('can call next with invalid operation', async () => {
      const req = {
        tenant: {
          id: tenantId,
        },
        user: {
          tenantId,
          id: 'test',
          name: 'tester',
          roles: [ServiceRoles.PdfGenerator],
        },
        body: {
          templateId: 'test',
          filename: 'test.pdf',
          fileType: 'my-test-type',
          data: {},
        },
        template: configuration.test,
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();

      repositoryMock.create.mockResolvedValueOnce({ id: 'job1' });
      fileServiceMock.typeExists.mockResolvedValueOnce(false);
      queueServiceMock.enqueue.mockResolvedValueOnce(null);
      const handler = generatePdf(
        serviceId,
        repositoryMock,
        eventServiceMock,
        fileServiceMock,
        queueServiceMock,
        loggerMock,
        serviceDirectoryMock
      );
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(InvalidOperationError));
      expect(res.send).not.toHaveBeenCalled();
    });

    it('can call next with unauthorized', async () => {
      const req = {
        tenant: {
          id: tenantId,
        },
        user: {
          tenantId,
          id: 'test',
          name: 'tester',
          roles: [],
        },
        body: {
          templateId: 'test',
          filename: 'test.pdf',
          data: {},
        },
        template: configuration.test,
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();

      const handler = generatePdf(
        serviceId,
        repositoryMock,
        eventServiceMock,
        fileServiceMock,
        queueServiceMock,
        loggerMock,
        serviceDirectoryMock
      );
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(UnauthorizedUserError));
      expect(res.send).not.toHaveBeenCalled();
    });
  });

  describe('getGeneratedFile', () => {
    it('can create handler', () => {
      const handler = getGeneratedFile(serviceId, repositoryMock);
      expect(handler).toBeTruthy();
    });

    it('can get file', async () => {
      const req = {
        tenant: {
          id: tenantId,
        },
        user: {
          tenantId,
          id: 'test',
          name: 'tester',
          roles: [ServiceRoles.PdfGenerator],
        },
        params: {
          jobId: 'job1',
        },
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();

      repositoryMock.get.mockResolvedValueOnce({ tenantId, id: req.params.jobId, status: 'completed' });
      const handler = getGeneratedFile(serviceId, repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ id: req.params.jobId, status: 'completed' }));
      expect(next).not.toHaveBeenCalled();
      expect(res.send.mock.calls[0][0]).toMatchSnapshot();
    });

    it('can call next with not found', async () => {
      const req = {
        tenant: {
          id: tenantId,
        },
        user: {
          tenantId,
          id: 'test',
          name: 'tester',
          roles: [ServiceRoles.PdfGenerator],
        },
        params: {
          jobId: 'job1',
        },
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();

      repositoryMock.get.mockResolvedValueOnce(null);
      const handler = getGeneratedFile(serviceId, repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(res.send).not.toHaveBeenCalled();
      expect(next).toHaveBeenCalledWith(expect.any(NotFoundError));
    });

    it('can call next with unauthorized', async () => {
      const req = {
        tenant: {
          id: tenantId,
        },
        user: {
          tenantId,
          id: 'test',
          name: 'tester',
          roles: [],
        },
        params: {
          jobId: 'job1',
        },
      };
      const res = {
        send: jest.fn(),
      };
      const next = jest.fn();

      repositoryMock.get.mockResolvedValueOnce({ tenantId, id: req.params.jobId, status: 'completed' });
      const handler = getGeneratedFile(serviceId, repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(res.send).not.toHaveBeenCalled();
      expect(next).toHaveBeenCalledWith(expect.any(UnauthorizedUserError));
    });
  });
});
