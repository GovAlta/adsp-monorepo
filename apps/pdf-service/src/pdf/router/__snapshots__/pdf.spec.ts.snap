// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pdf generatePdf can generate pdf 1`] = `
Object {
  "id": "job1",
  "result": null,
  "status": undefined,
  "urn": "urn:ads:platform:pdf-service:v1:/jobs/job1",
}
`;

exports[`pdf getGeneratedFile can get file 1`] = `
Object {
  "id": "job1",
  "result": null,
  "status": "completed",
  "urn": "urn:ads:platform:pdf-service:v1:/jobs/job1",
}
`;

exports[`pdf getTemplates can get templates 1`] = `
Array [
  Object {
    "description": "This is a test.",
    "id": "test",
    "name": "Test template",
    "template": "",
  },
]
`;
